src/addons/tilt.cpp

```diff
@@ -0,0 +1,296 @@
+											pinTiltLeftAnalogLeft,
+											pinTiltLeftAnalogRight,
+											pinTiltRightAnalogDown,
+											pinTiltRightAnalogUp,
+											pinTiltRightAnalogLeft,
+											pinTiltRightAnalogRight };
+
+	for (int i = 0; i < 11; i++) {
+		if (pinTilt[i] != (uint8_t)-1) {
+			gpio_init(pinTilt[i]);             // Initialize pin
+			gpio_set_dir(pinTilt[i], GPIO_IN); // Set as INPUT
+			gpio_pull_up(pinTilt[i]);          // Set as PULLUP
+		}
+	}
+
+	dDebLeftState = 0;
+	dDebRightState = 0;
+	tiltLeftState = 0;
+	tiltRightState = 0;
+
+	lastGPUD = DIRECTION_NONE;
+	lastGPLR = DIRECTION_NONE;
+
+	lastTiltUD = DIRECTION_NONE;
+	lastTiltLR = DIRECTION_NONE;
+
+	uint32_t now = getMillis();
+	for (int i = 0; i < 4; i++) {
+		dpadTime[i] = now;
+	}
+}
+
+void TiltInput::debounce()
+{
+	uint32_t now = getMillis();
+	Gamepad* gamepad = Storage::getInstance().GetGamepad();
+
+	for (int i = 0; i < 4; i++)
+	{
+		if ((dDebLeftState & dpadMasks[i]) != (tiltLeftState & dpadMasks[i]) && (now - dpadTime[i]) > gamepad->debounceMS)
+		{
+			dDebLeftState ^= dpadMasks[i];
+			dpadTime[i] = now;
+		}
+	}
+	tiltLeftState = dDebLeftState;
+
+	for (int i = 0; i < 4; i++)
+	{
+		if ((dDebRightState & dpadMasks[i]) != (tiltRightState & dpadMasks[i]) && (now - dpadTime[i]) > gamepad->debounceMS)
+		{
+			dDebRightState ^= dpadMasks[i];
+			dpadTime[i] = now;
+		}
+	}
+	tiltRightState = dDebRightState;
+}
+
+void TiltInput::preprocess()
+{
+	Gamepad* gamepad = Storage::getInstance().GetGamepad();
+
+	// Need to invert since we're using pullups
+	tiltLeftState = 0;
+	if (pinTiltLeftAnalogUp != (uint8_t)-1) {
+		tiltLeftState |= (!gpio_get(pinTiltLeftAnalogUp) ? gamepad->mapDpadUp->buttonMask : 0);
+	}
+	if (pinTiltLeftAnalogDown != (uint8_t)-1) {
+		tiltLeftState |= (!gpio_get(pinTiltLeftAnalogDown) ? gamepad->mapDpadDown->buttonMask : 0);
+	}
+	if (pinTiltLeftAnalogLeft != (uint8_t)-1) {
+		tiltLeftState |= (!gpio_get(pinTiltLeftAnalogLeft) ? gamepad->mapDpadLeft->buttonMask : 0);
+	}
+	if (pinTiltLeftAnalogRight != (uint8_t)-1) {
+		tiltLeftState |= (!gpio_get(pinTiltLeftAnalogRight) ? gamepad->mapDpadRight->buttonMask : 0);
+	}
+
+	tiltRightState = 0;
+	if (pinTiltRightAnalogUp != (uint8_t)-1) {
+		tiltRightState |= (!gpio_get(pinTiltRightAnalogUp) ? gamepad->mapDpadUp->buttonMask : 0);
+	}
+	if (pinTiltRightAnalogDown != (uint8_t)-1) {
+		tiltRightState |= (!gpio_get(pinTiltRightAnalogDown) ? gamepad->mapDpadDown->buttonMask : 0);
+	}
+	if (pinTiltRightAnalogLeft != (uint8_t)-1) {
+		tiltRightState |= (!gpio_get(pinTiltRightAnalogLeft) ? gamepad->mapDpadLeft->buttonMask : 0);
+	}
+	if (pinTiltRightAnalogRight != (uint8_t)-1) {
+		tiltRightState |= (!gpio_get(pinTiltRightAnalogRight) ? gamepad->mapDpadRight->buttonMask : 0);
+	}
+
+	// Debounce our directional pins
+	debounce();
+
+	// Convert gamepad from process() output to uint8 value
+	uint8_t gamepadState = gamepad->state.dpad;
+
+	gamepad->state.dpad = gamepadState;
+}
+
```

Comment:
I'm unsure why `TiltInput::preprocess()` is needed, and why it performs it has its own debouncing code before gamepad inputs are processed. Is there a specific reason for this? It seems this can be simplified by handling the tilt state logic in the `TiltInput::process()` function after GP2040-CE does its normal input `read()` and `process()` which will already handle debouncing.

Reply:
I added this because I don't fully understand GP2040-CE, I wrote the code referring to DualDirectional. Just to clarify, this game uses both the light and left analog sticks. Occasionally DPad is also used, but only in certain situations. If I could simplify it, I would.

Reply:
Thanks @smashBrosKanu for the quick response to this and all the feedback so far!

I'll to try and test this over the next day or two and see if there's some cleanup/optimization we can do around the input processing part of this addon.

Reply:
Thank you for your kind review. And I will learn a lot from you.